### PR TITLE
rewrite DataFrame.shift and Series.shift

### PR DESCRIPTION
**Pull request by RainFung**

### CODE REVIEW

**Comment by RainFung** [comment]
Wait to be code review.

**Comment by RainFung** [comment]
wait to be code reviewed.

File: databricks/koalas/series.py [hunk 1]
```diff
@@ -2650,6 +2650,60 @@ def _diff(self, periods, part_cols=()):
...
+            performance degradation. Avoid this method against very large dataset.
+
+        Parameters
+        ----------
+        periods : int
+            Number of periods to shift. Can be positive or negative.
+        fill_value : object, optional
+            The scalar value to use for newly introduced missing values.
+            The default depends on the dtype of self. For numeric data, np.nan is used.
+        Returns
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: Can we add one space before "Returns"?

File: databricks/koalas/series.py [hunk 2]
```diff
@@ -2650,6 +2650,60 @@ def _diff(self, periods, part_cols=()):
...
+        -------
+        Copy of input DataFrame, shifted.
+        Examples
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: Can we add one space before "Examples"?

File: databricks/koalas/series.py [hunk 3]
```diff
@@ -2650,6 +2650,60 @@ def _diff(self, periods, part_cols=()):
...
+        ...                    'Col3': [17, 27, 22, 37, 52]},
+        ...                   columns=['Col1', 'Col2', 'Col3'])
+        >>> df['Col1'].shift(periods=3)
+        0     NaN
+        1     NaN
+        2     NaN
+        3    10.0
+        4    20.0
+        Name: Col1, dtype: float64
+        >>> df['Col1'].shift(periods=3, fill_value=0)
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: Can we add one space before "df['Col1'].shift(periods=3, fill_value=0)"?

File: databricks/koalas/series.py [hunk 4]
```diff
@@ -2650,6 +2650,60 @@ def _diff(self, periods, part_cols=()):
...
+
+    def _shift(self, periods, fill_value, part_cols=()):
+        if len(self._internal.index_columns) == 0:
+            raise ValueError("Index must be set.")
+        if not isinstance(periods, int):
+            raise ValueError('periods should be an int')
+        window = Window.partitionBy(*part_cols).orderBy(self._internal.index_scols)\
+            .rowsBetween(-periods, -periods)
+        scol = F.when(F.lag(self._scol, periods).over(window).isNotNull(),
+                      F.lag(self._scol, periods).over(window)).otherwise(fill_value)
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: I would write it:

```python
lag_col = F.lag(self._scol, periods).over(window)
scol = F.when(lag_col.isNull(), fill_value).otherwise(lag_col)
```

File: databricks/koalas/frame.py [hunk 5]
```diff
@@ -2042,12 +2042,10 @@ def shift(self, periods=1, fill_value=None):
         2     0     0     0
         3    10    13    17
         4    20    23    27
-
         """
         applied = []
         for column in self._internal.data_columns:
             applied.append(self[column].shift(periods, fill_value))
-
```
Comment (HyukjinKwon) [inline databricks/koalas/frame.py]: Can you remove unrelated changes here?

**Comment by RainFung** [comment]
> Actually @RainFung, this feature is already implemented at `IndexOpsMixin`

emmm... sorry, i haven't find `shift` IndexOpsMixin. I'll check it next time.
The purpose of this PR is to create groupby.shift PR in the next step.

**Comment by HyukjinKwon** [comment]
Hm, I think you can use `shift` in `IndexOpsMixin` directly. `Series` inherits `IndexOpsMixin`.

File: databricks/koalas/series.py [hunk 6]
```diff
@@ -2650,6 +2650,62 @@ def _diff(self, periods, part_cols=()):
...
+        3    10
+        4    20
+        Name: Col1, dtype: int64
+        """
+        return self._shift(periods, fill_value)
+
+    def _shift(self, periods, fill_value, part_cols=()):
+        if not isinstance(periods, int):
+            raise ValueError('periods should be an int')
+        window = Window.partitionBy(*part_cols).orderBy(self._internal.index_scols)\
```
Comment (HyukjinKwon) [inline databricks/koalas/series.py]: Why don't you fix the one in `IndexOpsMixin` and reuse it in `GroupBy`?

Comment (RainFung) [reply databricks/koalas/series.py]: Fix it in new version.

File: databricks/koalas/base.py [hunk 7]
```diff
@@ -699,14 +699,15 @@ def shift(self, periods=1, fill_value=None):
         >>> df.index.shift(periods=3, fill_value=0)
         Int64Index([0, 0, 0, 0, 1], dtype='int64')
         """
+        return self._shift(periods, fill_value)
+
+    def _shift(self, periods, fill_value, part_cols=()):
         if not isinstance(periods, int):
-            raise ValueError('periods should be an int; however, got [%s]' % type(periods))
```
Comment (HyukjinKwon) [inline databricks/koalas/base.py]: Why did we remove this?

File: databricks/koalas/base.py [hunk 8]
```diff
@@ -699,14 +699,15 @@ def shift(self, periods=1, fill_value=None):
...
-        window = Window.orderBy(self._kdf._internal.index_scols).rowsBetween(-periods, -periods)
-        shifted_col = F.lag(col, periods).over(window)
-        col = F.when(
-            shifted_col.isNull() | F.isnan(shifted_col), fill_value
-        ).otherwise(shifted_col)
-
+        window = Window.partitionBy(*part_cols).orderBy(self._internal.index_scols)\
+            .rowsBetween(-periods, -periods)
+        lag_col = F.lag(col, periods).over(window)
+        col = F.when(lag_col.isNull(), fill_value).otherwise(lag_col)
```
Comment (HyukjinKwon) [inline databricks/koalas/base.py]: @RainFung, let's keep `F.isnan(shifted_col)`. In Spark, NULL and NaN are different. Adding it should be better to be robust.
